userspace/falco/falco_metrics.cpp

```diff
@@ -0,0 +1,114 @@
+*/
+
+#include "falco_metrics.h"
+
+#include "app/state.h"
+
+#include <libsinsp/sinsp.h>
+
+falco_metrics::falco_metrics(falco::app::state& state)
+{
+	falco_configuration::webserver_config webserver_config = state.config->m_webserver_config;
+	m_is_enabled = state.config->m_metrics_enabled && webserver_config.m_metrics_enabled;
+
+	if (m_is_enabled)
+	{
+		for (const auto& source_info: state.source_infos)
+		{
+			sinsp *source_inspector = source_info.inspector.get();
+			m_inspectors.push_back(source_inspector);
+			m_metrics_collectors.push_back(libs::metrics::libs_metrics_collector(source_inspector, state.config->m_metrics_flags));	
+		}
+	}
+}
+
+std::string falco_metrics::to_text() const
+{
+	if (!m_is_enabled)
+	{
+		return "";
+	}
+
+	static const char* all_driver_engines[] = {
+		BPF_ENGINE, KMOD_ENGINE, MODERN_BPF_ENGINE,
+		SOURCE_PLUGIN_ENGINE, NODRIVER_ENGINE, GVISOR_ENGINE };
+
+
+	libs::metrics::prometheus_metrics_converter converter;
+	std::string prometheus_text;
+
+	for (auto* inspector: m_inspectors)
+	{
+		for (size_t i = 0; i < sizeof(all_driver_engines) / sizeof(const char*); i++)
+		{
+			if (inspector->check_current_engine(all_driver_engines[i]))
+			{
+				prometheus_text += converter.convert_metric_to_text_prometheus("engine_name", "falcosecurity", "scap", {{"engine_name", all_driver_engines[i]}});
+				break;
+			}
+		}
+
+		const scap_agent_info* agent_info = inspector->get_agent_info();
+		const scap_machine_info* machine_info = inspector->get_machine_info();
+
+		libs::metrics::libs_metrics_collector libs_metrics_collector(inspector, 0);
+
+		prometheus_text += converter.convert_metric_to_text_prometheus("falco_version", "falcosecurity", "falco", {{"falco_version", FALCO_VERSION}});
+		prometheus_text += converter.convert_metric_to_text_prometheus("kernel_release", "falcosecurity", "falco", {{"kernel_release", agent_info->uname_r}});
+		prometheus_text += converter.convert_metric_to_text_prometheus("evt_hostname", "falcosecurity", "falco", {{"evt_hostname", machine_info->hostname}});
+
+		std::vector<metrics_v2> static_metrics;
+		static_metrics.push_back(libs_metrics_collector.new_metric("start_ts",
+											METRICS_V2_LIBBPF_STATS,
+											METRIC_VALUE_TYPE_U64,
+											METRIC_VALUE_UNIT_TIME_NS_COUNT,
+											METRIC_VALUE_METRIC_TYPE_MONOTONIC,
+											agent_info->start_ts_epoch));
+		static_metrics.push_back(libs_metrics_collector.new_metric("falco_host_boot_ts",
+											METRICS_V2_LIBBPF_STATS,
+											METRIC_VALUE_TYPE_U64,
+											METRIC_VALUE_UNIT_TIME_NS_COUNT,
+											METRIC_VALUE_METRIC_TYPE_MONOTONIC,
+											machine_info->boot_ts_epoch));
+		static_metrics.push_back(libs_metrics_collector.new_metric("falco_host_num_cpus",
+											METRICS_V2_LIBBPF_STATS,
+											METRIC_VALUE_TYPE_U64,
+											METRIC_VALUE_UNIT_TIME_NS_COUNT,
+											METRIC_VALUE_METRIC_TYPE_MONOTONIC,
+											machine_info->num_cpus));
+
+		for (auto metrics: static_metrics)
```

Comment:
> One thing that comes to mind: should we have a list to properly filter out all the metrics that should not be returned for Prometheus ?

Sure, at the same time I believe the ones we don't need for Prometheus are derived ones, hence we can also just skip creating them altogether. For example all of the libs metrics collector metrics are exposed, it's just a handful of derived ones we don't need for Prometheus, for instance the event rates or drop percentages.

Reply:
I might be missing something, aren't the missing metrics only available to the stats writer ?
In which case, it seems that there's a need for a deeper refactoring to make the information available in the webserver as well.

Reply:
`evt_source` -> available

`outputs_queue_num_drops` -> this is part of `falco_outputs` and not unique to stats_writer, search for method get_outputs_queue_num_drops, aka just need to make `falco_outputs` accessible, should be possible right?

`num_evts` -> this is sinsp number of events not to be confused with n_evts that are the scap number of events, it's counted within `do_inspect`, mind checking what would need to be done to make this accessible?

duration_sec -> calculate just like done in the stats writer - we have everything `(uint64_t)((now - agent_info->start_ts_epoch) / ONE_SECOND_IN_NS);`

Reply:
`outputs_queue_num_drops`: done
`duration_sec`: done

For `evt_source`, from a metrics point of view, it's a bit more nebular. There should be `syscall` available by default plus anything that configured plugin provides, so there might be more than one. Do you mean that the metrics should return an entry for each event source from an inspector ?

for `num_events` I am still looking for an elegant solution to retrieve or set the value.

Reply:
Thanks @sgaist!

Re num_evts I suspected that it may be a bit more tricky.

Re evt_source agreed and it's also one open discussion item since right now you loop over the event sources vs in the outputs rule metrics we do not. I would say right now metrics caters more to one event source only. Something we should improve and disentangle in the future ... out of scope for this PR. CC @jasondellaluce since if I recall correctly you asked to add it back then. For now I would add it for parity with the existing metrics. Lastly maybe we still have time for Falco 0.38.0 to address the event source for metrics across the board in a follow up PR.

Reply:
As a temporary solution, I am publishing all the sources for each inspector.

I am wondering, from an architecture point of view, if the metrics and stats data should be stored in a common "model" and then the writers and metrics backend would be views that get their data from that model to do their job.

Reply:
@sgaist +1 a common metrics model makes the most sense. Re metrics for the different sources ... we need to rethink this more as for Prometheus we can surely only host one metrics view. For now let's accept that it's only working for syscalls source well. Most metrics are for the syscalls event source in the first place. For the case where we solely have a plugin source some metrics (CPU usage) are right now slightly broken because of a scap refactor that introduced a regression.